src/claimtrie.cpp

```diff
@@ -312,57 +374,72 @@ void completeHash(uint256& partialHash, const std::string& key, std::size_t to)
             .Finalize(partialHash.begin());
 }
 
-bool recursiveCheckConsistency(CClaimTrie::const_iterator& it, int minimumHeight, std::string& failed)
+template <typename T>
+using iCbType = std::function<void(T&)>;
+
+template <typename TIterator>
+uint256 recursiveMerkleHash(TIterator& it, const iCbType<TIterator>& process, const iCbType<TIterator>& verify = {})
 {
     std::vector<uint8_t> vchToHash;
-
     const auto pos = it.key().size();
-    auto children = it.children();
-    for (auto& child : children) {
-        if (!recursiveCheckConsistency(child, minimumHeight, failed))
-            return false;
+    for (auto& child : it.children()) {
+        process(child);
         auto& key = child.key();
         auto hash = child->hash;
-        vchToHash.push_back(key[pos]);
         completeHash(hash, key, pos);
+        vchToHash.push_back(key[pos]);
         vchToHash.insert(vchToHash.end(), hash.begin(), hash.end());
     }
 
     CClaimValue claim;
     if (it->getBestClaim(claim)) {
-        if (it->nHeightOfLastTakeover < minimumHeight) {
-            LogPrintf("\nInvalid takeover height for %s\n", it.key());
-            failed = it.key();
-            return false;
-        }
         uint256 valueHash = getValueHash(claim.outPoint, it->nHeightOfLastTakeover);
         vchToHash.insert(vchToHash.end(), valueHash.begin(), valueHash.end());
-    } else if (children.empty()) {
-        // we don't allow a situation of no children and no claims; no empty leaf nodes allowed
-        LogPrintf("\nInvalid empty node for %s\n", it.key());
-        failed = it.key();
-        return false;
+    } else if (verify) {
+        verify(it);
     }
 
-    auto calculatedHash = computeHash(vchToHash);
-    auto matched = calculatedHash == it->hash;
-    if (!matched) {
-        LogPrintf("\nComputed hash doesn't match stored hash for %s\n", it.key());
-        failed = it.key();
+    return Hash(vchToHash.begin(), vchToHash.end());
+}
+
+bool recursiveCheckConsistency(CClaimTrie::const_iterator& it, std::string& failed)
+{
+    struct CRecursiveBreak{};
+
+    using iterator = CClaimTrie::const_iterator;
+    iCbType<iterator> verify = [&failed](iterator& it) {
+        if (!it.hasChildren()) {
+            // we don't allow a situation of no children and no claims; no empty leaf nodes allowed
+            failed = it.key();
+            throw CRecursiveBreak();
+        }
+    };
+
+    iCbType<iterator> process = [&failed, &process, &verify](iterator& it) {
+        if (it->hash != recursiveMerkleHash(it, process, verify)) {
+            failed = it.key();
+            throw CRecursiveBreak();
+        }
+    };
+
+    try {
```

Comment:
A recursive break is not an exceptional condition. This should use return values instead. try/catch should be extremely rare and the catch should never be the expected scenario.

Reply:
It is https://github.com/lbryio/lbrycrd/pull/291/commits/35a987bf2b7bf7b4d4f5b08297c5a45a4d467492#diff-f8129adb47c1b30eac7d2795b17883edR427
That's easy way to stop recursion, it does not affect other exceptions.

Reply:
How is it easier than using a return value?

Reply:
It breaks the stack without unwinding, i.e. it does not pop all stack pushes and the stack is unwind when catch appears. https://en.wikipedia.org/wiki/Call_stack#Unwinding